Update ubi9 and add ubi9-based developer image

dkwon17 opened the pull request:
Part of https://github.com/eclipse-che/che/issues/23034

To test:

1. Set env vars:

```
BASE_IMAGE=quay.io/<username>/<repo>:ubi9
UDI=quay.io/<username/<repo>:udi
```

2. Build and push the base image

```
cd base/ubi9
DOCKER_BUILDKIT=1 docker image build --progress=plain -t $BASE_IMAGE .
docker push $BASE_IMAGE
```

3. Build and push the udi image  
    First, replace line 4 of the udi dockerfile [here](https://github.com/dkwon17/developer-images/blob/b4140ecc24b8144f3a858a6be39f25202d362e84/universal/ubi9/Dockerfile#L4) so that the base image matches the base image built from step 2.

Then, run the following:

```
cd universal/ubi9
DOCKER_BUILDKIT=1 docker image build --progress=plain -t $UDI .
docker push $UDI
```

I've built and pushed my own images for testing here:
```
quay.io/dkwon17/base-developer-image:ubi9-latest
quay.io/dkwon17/universal-developer-image:ubi9-latest
```

4. Test by creating workspaces for the following cases

<details><summary>

#### Case 1

</summary>

The workspace starts with the base image:
```
<CHE_HOST>/#https://github.com/dkwon17/empty?image=<BASE_IMAGE>
```

</details>

<details><summary>

#### Case 2

</summary>


The workspace starts with the UDI image `<CHE_HOST>/#https://github.com/dkwon17/quarkus-api-example/tree/per-workspace?image=<UDI>`

Next, verify that the following command runs without any errors:

```
podman info
```

Run the `Package` and `Build image` devfile tasks to verify that podman image builds are also working.

Lastly, test kubedock by adding the `KUBEDOCK_ENABLED=true` env variable in the tooling container:

```
oc patch dw <workspace-name> --type='json' -p='[{"op": "add", "path": "/spec/template/components/0/container/env", "value": [{"name": "KUBEDOCK_ENABLED", "value": "true"}]}]' -n <namespace>
```

and by running the following when the workspace restarts:

```
$ podman run hello-world
!... Hello Podman World ...!

         .--"--.           
       / -     - \         
      / (O)   (O) \        
   ~~~| -=(,Y,)=- |         
    .---. /`  \   |~~      
 ~/  o  o \~~~~.----. ~~   
  | =(X)= |~  / (O (O) \   
   ~~~~~~~  ~| =(Y_)=-  |   
  ~~~~    ~~~|   U      |~~ 

Project:   https://github.com/containers/podman
Website:   https://podman.io
Desktop:   https://podman-desktop.io
Documents: https://docs.podman.io
YouTube:   https://youtube.com/@Podman
X/Twitter: @Podman_io
Mastodon:  @Podman_io@fosstodon.org
```
Additionally check that extensions from .vscode/extensions are installed:

![image](https://github.com/user-attachments/assets/2a68db47-cd64-4f9c-ab7c-5522b06199bf)


</details>



<details><summary>

#### Case 3

</summary>

The workspace starts with the UDI image and podman build works.

After testing case 2, run the `devfile: Package` and the `devfile: Build Image` devfile tasks to build the container image:
```
...
STEP 9/10: USER 185
--> b2fc97577056
STEP 10/10: ENTRYPOINT [ "java", "-jar", "/deployments/quarkus-run.jar" ]
COMMIT image-registry.openshift-image-registry.svc:5000/openshift/quarkus-api-example
--> c44f4cfddabb
Successfully tagged image-registry.openshift-image-registry.svc:5000/openshift/quarkus-api-example:latest
c44f4cfddabb7817beec39951d1ceaabd4c575dbdf5a8e2b9a68ae012a383652
 *  Terminal will be reused by tasks, press any key to close it. 
```

Also, verify that the `devfile: Start Development Mode` command also runs without issues:
```
...
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ quarkus-api-example ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- quarkus-maven-plugin:2.6.3.Final:dev (default-cli) @ quarkus-api-example ---
[INFO] Invoking org.apache.maven.plugins:maven-resources-plugin:2.6:testResources) @ quarkus-api-example
[e/src/main/kubernetes.ng for existing resources in: /projects/quarkus-api-exampl
__  ____  __  _____   ___  __ ____  ______ /projects/quarkus-api-example/src/test/resources
2024-11-13 04:34:09,161 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [agroal, cdi, hibernate-orm, hibernate-orm-panache, jdbc-postgresql, kubernetes, kubernetes-client, narayana-jta, resteasy, resteasy-jackson, smallrye-context-propagation, smallrye-openapi, swagger-ui, vertx]
--
Tests paused
Press [r] to resume testing, [o] Toggle test output, [h] for more options>
```
</details>

dkwon17 commented on base/ubi9/Dockerfile:
```diff
@@ -19,17 +19,26 @@ LABEL io.openshift.expose-services=""
     TEMP_DIR="$(mktemp -d)"; \
     cd "${TEMP_DIR}"; \
-    GH_VERSION="2.23.0"; \
+    GH_VERSION="2.45.0"; \
```
This was done to match the GH version in the ubi8 image: https://github.com/devfile/developer-images/blob/0eb879a7fccd12da128f89b99a986292a0b9091a/base/ubi8/Dockerfile#L42

dkwon17 commented on universal/ubi9/Dockerfile:
```diff
@@ -0,0 +1,430 @@
+RUN pip install pylint yq
+
+# PHP
+ENV PHP_VERSION=8.1
```
I want to point out that for the ubi8 based udi, the PHP_VERSION is 7.4:
https://github.com/devfile/developer-images/blob/0eb879a7fccd12da128f89b99a986292a0b9091a/universal/ubi8/Dockerfile#L140

But for ubi9, the supported streams by default are:
```
Name Stream Profiles               Summary               
php  8.1    common, devel, minimal PHP scripting language
php  8.2    common, devel, minimal PHP scripting language
```

so I've updated the PHP_VERSION to 8.1.

Hello @svor , is there a strong reason for the UDI to stay on PHP 7.4?

svor replied:
hi @dkwon17, I don't it's a problem having PHP 8.1 instead of 7.4

dkwon17 replied:
Thank you for the clarification. In my latest commits, I've updated the version to 8.2 which is the latest available stream

svor reviewed (commented):
@dkwon17  great work! All cases from the description work as expected. A few moments:
- In the description (step 2 and 3) should be `cd base/ubi9` and `cd universal/ubi9` 
- Maybe we should add an information about ubi9 and udi9 into README.md and CONTRIBUTING.md files as well?
- Should we have a GH action for udi9 like `Empty workspace smoke test on udi9`?

dkwon17 commented on base/ubi9/Dockerfile:
```diff
@@ -88,16 +97,71 @@ RUN \
+   chown -R 10001 "${HOME}"/.config
+
+# Add kubedock
+ENV KUBEDOCK_VERSION 0.15.5
```
Sounds good, updated to 0.17.0

dkwon17 commented:
@svor thank you for the review!

> In the description (step 2 and 3) should be cd base/ubi9 and cd universal/ubi9
Fixed

> Maybe we should add an information about ubi9 and udi9 into README.md and CONTRIBUTING.md files as well?
I've added a commit to change some mentions of ubi8 to ubi9. Since we are planning to deprecate ubi8: https://github.com/eclipse-che/che/issues/23034 I plan to make another PR to be merged after this current PR, to remove the ubi8 code and update the rest of the readme

> Should we have a GH action for udi9 like Empty workspace smoke test on udi9?
I've added a new workflow for this in my latest commit

svor commented on .github/workflows/empty-worksapce-smoke-test-on-minikube-ubi9.yaml:
```diff
@@ -0,0 +1,116 @@
+#
+# Copyright (c) 2019-2021 Red Hat, Inc.
```
```suggestion
# Copyright (c) 2019-2024 Red Hat, Inc.
```

svor commented on .github/workflows/empty-worksapce-smoke-test-on-minikube-ubi9.yaml:
```diff
@@ -0,0 +1,116 @@
+      - .devfile.yaml
+      - LICENSE
+      - '.rebase/*'
+      - 'base/ubi9/**'
```
base/ubi8/** and /universal/ubi8/** should be ignored and .rebase/* can be removed (it doesn't exist).
Also in `empty-worksapce-smoke-test-on-minikube-ubi8.yaml` you can extend `paths-ignore` by adding 'universal/ubi9/**'

svor reviewed (approved):
Added a few comments, but in general changes look good!

dkwon17 commented:
Thank you @svor , I updated the PR

AObuchow commented on universal/ubi9/Dockerfile:
```diff
@@ -0,0 +1,430 @@
+# Cloud
+
+# oc client
+ENV OC_VERSION=4.15
```
Maybe we should update OC to the latest 4.17 (or 4.16 which is the first version of oc built specifically for RHEL 9 containers?). Though perhaps we're trying to keep as close of a 1-1 compatibility between the UBI8 and UBI9 based UDI images.

dkwon17 replied:
Let's keep it as is for this PR, and potentially upgrade in a different PR

AObuchow commented on universal/ubi9/Dockerfile:
```diff
@@ -0,0 +1,430 @@
+
+LABEL com.redhat.component="devfile-universal-container"
+LABEL name="devfile/universal-developer-image"
+LABEL version="ubi8"
```
This should be updated to `version="ubi9"`

AObuchow commented on universal/ubi9/Dockerfile:
```diff
@@ -0,0 +1,430 @@
+LABEL maintainer="Red Hat, Inc."
+
+LABEL com.redhat.component="devfile-universal-container"
+LABEL name="devfile/universal-developer-image"
```
Should we maybe differentiate the labels between the UBI9-based UDI and the UB8-based UDI? E.g. devfile/universal-developer-image-ubi9? Though I suppose this is already accomplished using different image tags?

AObuchow replied:
On second though, I don't think this is necessary since UBI 8 will be deprecated as mentioned [here](https://github.com/devfile/developer-images/pull/190#discussion_r1848855421)

AObuchow commented on universal/ubi9/Dockerfile:
```diff
@@ -0,0 +1,430 @@
+# Configure container engine
+COPY --chown=0:0 containers.conf /etc/containers/containers.conf
+
+ENV K8S_VERSION=1.28
```
Note, if we update OC we should also update to the corresponding k8s (kubectl) version that's bundled with OC.

AObuchow reviewed (commented):
Left some very minor comments but overall this looks great to me, awesome work @dkwon17 :)

dkwon17 commented on base/ubi9/.stow-local-ignore:
Yes, UBI8 will be deprecated: https://github.com/eclipse-che/che/issues/23034

AObuchow replied:
Sounds good, in that case there's no point in sharing the common files between the UBI8-based UDI & UBI9-based UDI.

dkwon17 commented on base/ubi9/Dockerfile:
```diff
@@ -19,17 +19,26 @@ LABEL io.openshift.expose-services=""
+ENV HOME=/home/tooling
+RUN mkdir -p /home/tooling/
+
 # Removed because of vulnerabilities: git-lfs
```
In my latest commits, I've added git-lfs to maintain some parity with ubi8

dkwon17 commented on base/ubi9/Dockerfile:
```diff
@@ -19,17 +19,26 @@ LABEL io.openshift.expose-services=""
 # Removed because of vulnerabilities: git-lfs
-RUN dnf install -y diffutils git iproute jq less lsof man nano procps \
-    perl-Digest-SHA net-tools openssh-clients rsync socat sudo time vim wget zip && \
+RUN dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm && \
```
Good idea, updated in my commit: d7ac8449c8c81138f33c0d736b02d379f904cf39

dkwon17 commented on base/ubi9/Dockerfile:
```diff
@@ -88,16 +97,71 @@ RUN \
+ENV ORIGINAL_PODMAN_PATH=/usr/bin/podman.orig
+COPY --chown=0:0 podman-wrapper.sh "${PODMAN_WRAPPER_PATH}"
+
+COPY --chown=0:0 podman-wrapper.sh /usr/bin/podman.wrapper
```
Updated to remove the second COPY in d7ac8449c8c81138f33c0d736b02d379f904cf39

AObuchow reviewed (approved):
Looks good to me, amazing work @dkwon17 :)

AObuchow commented:
Additional steps that should follow this PR (or could be included in it):
- Update the GH Action to build and push the UBI9-based UDI, as we're only building the [base image](https://github.com/devfile/developer-images/blob/main/.github/workflows/ubi9-build.yaml#L39-L48). I think we should be able to copy most of the [logic](https://github.com/devfile/developer-images/blob/0eb879a7fccd12da128f89b99a986292a0b9091a/.github/workflows/ubi8-build.yaml#L61-L111) used for the UBI8-based UDI
- Update the README ([here](https://github.com/devfile/developer-images/blob/0eb879a7fccd12da128f89b99a986292a0b9091a/README.md?plain=1#L13-L17) and [here](https://github.com/devfile/developer-images/blob/0eb879a7fccd12da128f89b99a986292a0b9091a/README.md?plain=1#L92-L96)) to reference the UBI9-based images once they're hosted on quay.io,
